javalib/src/main/scala/java/util/ArrayList.scala

```diff
@@ -14,80 +14,187 @@ package java.util
 
 import java.lang.Cloneable
 import java.lang.Utils._
+import java.util.ScalaOps._
 
 import scala.scalajs._
+import scala.scalajs.LinkingInfo.isWebAssembly
 
-class ArrayList[E] private (private[ArrayList] val inner: js.Array[E])
+class ArrayList[E] private (innerInit: AnyRef, private var _size: Int)
     extends AbstractList[E] with RandomAccess with Cloneable with Serializable {
   self =>
 
+  /* This class has two different implementations for handling the
+   * internal data storage, depending on whether we are on Wasm or JS.
+   * On JS, we utilize `js.Array`. On Wasm, for performance reasons,
+   * we avoid JS interop and use a scala.Array.
+   * The `_size` field (unused in JS) keeps track of the effective size
+   * of the underlying Array for the Wasm implementation.
+   */
+
+  private val innerJS =
+    if (isWebAssembly) null
+    else innerInit.asInstanceOf[js.Array[E]]
+
+  private var innerWasm =
+    if (!isWebAssembly) null
+    else innerInit.asInstanceOf[Array[AnyRef]]
+
   def this(initialCapacity: Int) = {
-    this(new js.Array[E])
-    if (initialCapacity < 0)
-      throw new IllegalArgumentException
+    this(
+      {
+        if (initialCapacity < 0)
+          throw new IllegalArgumentException
+        if (isWebAssembly) new Array[Any](initialCapacity)
+        else new js.Array[E]
+      },
+      0
+    )
   }
 
   def this() =
-    this(new js.Array[E])
+    this(
+      if (isWebAssembly) new Array[AnyRef](16)
+      else new js.Array[E],
+      0
+    )
 
   def this(c: Collection[_ <: E]) = {
-    this()
+    this(
+      if (isWebAssembly) new Array[AnyRef](c.size())
+      else new js.Array[E],
+      0
+    )
     addAll(c)
   }
 
   def trimToSize(): Unit = {
-    // We ignore this as js.Array doesn't support explicit pre-allocation
+    if (isWebAssembly)
+      resizeTo(size())
+    // We ignore this in JS as js.Array doesn't support explicit pre-allocation
   }
 
   def ensureCapacity(minCapacity: Int): Unit = {
-    // We ignore this as js.Array doesn't support explicit pre-allocation
+    if (isWebAssembly) {
+      if (innerWasm.length < minCapacity)
+        resizeTo(minCapacity)
```

Comment:
Should we round up to the next power of 2 here?

Reply:
Sounds good :+1:

https://github.com/scala-js/scala-js/pull/5162/files#diff-de47e70080341dc21a1ea1f68a02da92b8a6f85065317b155b2f79d12fbe7164R69-R79